crates/proto/src/clients/mod.rs

```diff
@@ -0,0 +1,633 @@
+            endpoint: endpoint.into(),
+            timeout: None,
+        }
+    }
+
+    /// Creates a new remote client manager with an optional custom timeout.
+    ///
+    /// If `timeout` is `None`, uses the default 10-second timeout.
+    /// If `timeout` is `Some(duration)`, uses the specified timeout.
+    pub fn with_timeout(endpoint: impl Into<String>, timeout: Option<Duration>) -> Self {
+        Self {
+            client: Arc::new(tokio::sync::Mutex::new(None)),
+            endpoint: endpoint.into(),
+            timeout,
+        }
+    }
+
+    /// Connects to the remote service using a provided factory function.
+    ///
+    /// This method handles the connection logic, including:
+    /// - Checking if already connected (returns existing client)
+    /// - Creating a configured `ClientBuilder` with TLS and timeout
+    /// - Calling the factory function to create the specific client type
+    /// - Storing the client for reuse
+    ///
+    /// # Arguments
+    /// * `factory` - Async function that creates the client given an endpoint and configured
+    ///   `ClientBuilder`
+    ///
+    /// # Returns
+    /// A clone of the connected client
+    pub async fn connect_with<F, Fut, E>(&self, factory: F) -> Result<C, E>
+    where
+        F: FnOnce(String, ClientBuilder) -> Fut,
+        Fut: std::future::Future<Output = Result<C, E>>,
+    {
+        let mut client_guard = self.client.lock().await;
+
+        // Return existing client if already connected
+        if let Some(client) = client_guard.as_ref() {
+            return Ok(client.clone());
+        }
+
+        // Create configured builder
+        let mut builder = ClientBuilder::new().with_tls();
+
+        // Apply custom timeout if specified (`ClientBuilder` defaults to 10 seconds)
+        if let Some(timeout) = self.timeout {
+            builder = builder.with_timeout(timeout);
+        }
+
+        // Create the client using the factory
+        let new_client = factory(self.endpoint.clone(), builder).await?;
+
+        // Store and return the client
+        *client_guard = Some(new_client.clone());
+        Ok(new_client)
+    }
+
+    /// Gets a clone of the current client if connected.
+    ///
+    /// Returns `None` if no connection has been established yet.
+    pub async fn get_client(&self) -> Option<C> {
+        self.client.lock().await.clone()
+    }
+
+    /// Gets the endpoint URL.
+    pub fn endpoint(&self) -> &str {
+        &self.endpoint
+    }
+
+    /// Gets the configured timeout.
+    pub fn timeout(&self) -> Option<Duration> {
+        self.timeout
+    }
+}
+
+// WASM-specific factory helper for remote prover clients
+#[cfg(target_arch = "wasm32")]
+impl<C> RemoteClientManager<C>
+where
+    C: Clone,
+{
+    /// Connects using a WASM-compatible client factory.
+    ///
+    /// This is a convenience method for WASM targets that don't use `ClientBuilder`.
+    pub async fn connect_wasm_with<F, E>(&self, factory: F) -> Result<C, E>
```

Comment:
Is there any big difference between this and `connect_with()`? It seems that the `wasm-32`-specific changes are related to what happens within the passed function (basically, defining the tonic transport/channel), right? If so maybe this can be generic over the transport directly and the user could choose what to do there